### PR TITLE
allocator/mmaprototype: add repair orchestration and AddVoter

### PR DESCRIPTION
**Pull request by tbg**
Second PR in the MMA repair productionization stack (see #164658 for the
prototype and `how-to-productionize.md` for the full plan). Builds on the
repair foundation from #165413.

This PR adds:

- **Repair orchestration loop**: `repair()` on `rebalanceEnv` iterates
  `repairRanges` in priority order, filters to ranges where the local store
  is leaseholder, and dispatches to per-action repair functions. Integrated
  into `ComputeChanges` via the `IncludeRepair` option on `ChangeOptions`.

- **repairAddVoter**: the first concrete repair action. Two code paths:
  (1) promote an existing non-voter to voter when one satisfies voter
  constraints; (2) add a new voter on a constraint-satisfying,
  diversity-maximizing store.

- **Repair helpers**: `pickStoreByDiversity` (diversity-based store selection
  with reservoir sampling for tie-breaking), `filterAddCandidates` (filters
  to ready stores not already hosting a replica at the node level),
  `enactRepair` (records pending change and appends to change list),
  `isLeaseholderOnStore`, `replicaStateForStore`.

- **DSL test infrastructure**: `repair` command that creates a `rebalanceEnv`
  and runs `repair()`, with deterministic random seed for reproducible output.

The remaining 11 repair actions (RemoveVoter, ReplaceDeadVoter, etc.) are
wired as stubs logging "not yet implemented" — they come in later PRs.

### Commits

1. **add repair orchestration loop** — infrastructure: `repair()` loop,
   `IncludeRepair` wiring, `originMMARepair` enum, DSL `repair` command.
   No actions implemented (all hit default "not yet implemented" log).

2. **implement repairAddVoter with non-voter promotion** — first action +
   all helpers: `enactRepair`, `filterAddCandidates`, `pickStoreByDiversity`,
   `repairAddVoter`, `promoteNonVoterToVoter`. Two new DSL tests.

Stacked on #165413. Informs #164658.

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
> @tbg reviewed 18 files and all commit messages, and made 16 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/165423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained.

___
*[`-- commits` line 182 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSUgM22JcTuqEfG_aF:-OnSUgM30y--O7AV-jUP:b-7sftw6):*
Can you update the commit message to contrast this with how the legacy allocator works and to highlight any differences?
Interactively let me engage first before pushing the suggested update.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 306 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSQ1l7BRSYNgP-ZLRq:-OnSQ1l7BRSYNgP-ZLRr:b2ibfcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c828bdf9e0b2c309d661bd14d1307e4c20bf4e6d/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L306)):*
> ```Go
> ) []ExternalRangeChange {
> 	re.mmaid++
> 	ctx = logtags.AddTag(ctx, "mmaid", re.mmaid)
> ```

now we have more than one place that does this. Couldn't this be lifted to the caller so that we have a `ctx` that both repair and rebalance can use that already has the mmaid in it? Do this as a new commit in the end (i.e won't be squashed).
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 309 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSQQMw7Sizho-oj4FN:-OnSQQMx63V3jjGPUlId:b-nh7tnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c828bdf9e0b2c309d661bd14d1307e4c20bf4e6d/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L309)):*
> ```Go
> 
> 	// Iterate repair actions in priority order (lower enum = higher priority).
> 	for action := FinalizeAtomicReplicationChange; action < NoRepairNeeded; action++ {
> ```

- can you start at `1` (so that we don't think there's anything special about FinalizeAtomicReplicationChange)
- mention that 0 is not a valid action by design, in a comment (so nobody wonders if there's an off by one here)
- make a const numActions and make sure it doesn't rot. One way to do this is a unit test that validates that `Action(numActions)` still stringifies properly but `numActions+1` doesn't. You might be able to think of a better way too.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 314 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSRLX4-B3NAyy6S0g_:-OnSRLX4-B3NAyy6S0ga:bvz6p6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c828bdf9e0b2c309d661bd14d1307e4c20bf4e6d/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L314)):*
> ```Go
> 			continue
> 		}
> 		// Sort range IDs for deterministic iteration order.
> ```

See https://github.com/cockroachdb/cockroach/pull/165284 - this isn't merged but add a sync.Pool for this following that pattern regardless, so that we don't allocate slices in the common case.

I'm also worried by this determinism. We don't attempt to be entirely random, but we should also not be entirely not random. Keep the sorting, but then make `range ids {` iteration order deterministically random (using the rng on rebalanceEnv).
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 298 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSSw6UBGhCcmYnsHM_:-OnSSw6UBGhCcmYnsHMa:b-wpoo18) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L298)):*
> ```Go
> // replicaStateForStore returns the ReplicaState of the replica on the given
> // store, and whether it was found.
> func replicaStateForStore(rs *rangeState, storeID roachpb.StoreID) (ReplicaState, bool) {
> ```

can you make this a method on `rangeState` (and pick a good location)?
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 318 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSTAiU1B4mSadCaBFS:-OnSTAiU1B4mSadCaBFT:b-k5cev9) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L318)):*
> ```Go
> }
> 
> // filterAddCandidates filters candidateStores down to stores that are ready
> ```

For performance reasons, I would like the returned storeSet to be backed by the same memory as the incoming candidateStores set. This should be part of the commented contract and the caller needs to be careful when using pooled memory. See https://github.com/cockroachdb/cockroach/pull/165284 for examples of this pattern.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 322 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSTnd79K9QfhU0w832:-OnSTnd79K9QfhU0w833:b-5o037o) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L322)):*
> ```Go
> // range at the node level. excludeStoreID, if non-zero, is excluded from the
> // existing-replica set (used when a replica on that store is being
> // concurrently removed as part of the same change).
> ```

add: , such as during non-voter promotions
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSUPy29Y24T4R6TmqX:-OnSUPy29Y24T4R6TmqY:b-2cqakf) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L374)):*
> ```Go
> // for diversity scoring: voterLocalityTiers for voter operations,
> // replicaLocalityTiers for non-voter operations.
> func (re *rebalanceEnv) pickStoreByDiversity(
> ```

Does rebalancing (not repair) have an equivalent of this or does it not care about diversity yet?
This is a discussion item.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 404 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSUafC50nCS5o1-OW7:-OnSUafC50nCS5o1-OW8:b-sgmgl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L404)):*
> ```Go
> 
> // repairAddVoter attempts to add a voter to an under-replicated range.
> // It follows the decision tree from constraint.go: first try to promote a
> ```

the reference to `constraint.go` isn't going to age well, remove it.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 441 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSV-ez0x-fohJEpo4C:-OnSV-ez0x-fohJEpo4D:b7bsnxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L441)):*
> ```Go
> 	re.constraintMatcher.constrainStoresForExpr(constrDisj, &candidateStores)
> 
> 	validCandidates := re.filterAddCandidates(ctx, rs, candidateStores, 0)
> ```

use a local const for the `0`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 443 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSV62_2koMzHA1FzAb:-OnSV62_2koMzHA1FzAc:blgwyhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L443)):*
> ```Go
> 	validCandidates := re.filterAddCandidates(ctx, rs, candidateStores, 0)
> 	if len(validCandidates) == 0 {
> 		log.KvDistribution.Warningf(ctx,
> ```

Is this really a warning? I'd see this as a verbosity 1 Infof. This can happen routinely, and to many ranges, during outages or when constraints are misconfigured. Yes, worth changing, but hardly worth logging at high density or Warning.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 465 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSVU0v5MX6pcbgesMl:-OnSVU0v5MX6pcbgesMm:biveyk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L465)):*
> ```Go
> 	rangeChange := MakePendingRangeChange(rangeID, []ReplicaChange{addChange})
> 	if err := re.preCheckOnApplyReplicaChanges(rangeChange); err != nil {
> 		log.KvDistribution.Warningf(ctx,
> ```

we don't expect to hit this, right? Then Warningf is ok but please check that this can't be routinely hit (and possibly for many ranges in one go).
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 470 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSV_Q93kEYsWRslw90:-OnSV_Q93kEYsWRslw91:bufz6qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L470)):*
> ```Go
> 	}
> 	re.enactRepair(ctx, localStoreID, rangeChange)
> 	log.KvDistribution.Infof(ctx,
> ```

we don't want to log at Info for routine stuff except in the aggregate. Put this behind verbosity 1.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 490 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSVjErB27YH-G-4AyD:-OnSVjErB27YH-G-4AyE:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L490)):*
> ```Go
> 		(*existingReplicaLocalities).getScoreChangeForNewReplica)
> 	if bestStoreID == 0 {
> 		log.KvDistribution.Warningf(ctx,
> ```

ditto
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 498 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSVkGz6zqbeZraxpHP:-OnSVkGz6zqbeZraxpHQ:b-zcdnsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L498)):*
> ```Go
> 	prevState, found := replicaStateForStore(rs, bestStoreID)
> 	if !found {
> 		log.KvDistribution.Warningf(ctx,
> ```

This can remain a warning, right? Since we went down this path and so there really ought to be a non-voter we could promote and in particular its store should be known?
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 522 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnSVx3gDERIM9cIbToI:-OnSVx3gDERIM9cIbToJ:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/78c22d3ae9fe1cebabd8c0e1c8a0de8019036ac7/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L522)):*
> ```Go
> 	}
> 	re.enactRepair(ctx, localStoreID, rangeChange)
> 	log.KvDistribution.Infof(ctx,
> ```

ditto

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+			if len(m) == 0 {
+				delete(cs.repairRanges, oldAction)
+			}
+		} else if buildutil.CrdbTestBuild {
+			panic(errors.AssertionFailedf(
+				"repairRanges missing bucket for action %s on r%d", oldAction, rangeID))
+		}
+	}
+	// Add to new bucket.
+	if newAction != NoRepairNeeded && newAction != RepairPending {
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: /review-crdb(suggestion): `RepairSkipped` ranges (quorum loss, nil config) are added to `repairRanges` here but should be excluded like `RepairPending`. When `repair()` iterates, it hits the `default` case and logs `"repair action RepairSkipped for rN not yet implemented"` at Info level — misleading (these are intentionally skipped, not unimplemented) and potentially noisy in stressed clusters.

```suggestion
	if newAction != NoRepairNeeded && newAction != RepairPending && newAction != RepairSkipped {
```

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+		return
+	}
+
+	// Pick the target with the best voter diversity score.
+	bestStoreID := re.pickStoreByDiversity(
+		validCandidates, rs.constraints.voterLocalityTiers,
+		(*existingReplicaLocalities).getScoreChangeForNewReplica)
+
+	// Create the pending change.
+	targetSS := re.stores[bestStoreID]
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go:460]: /review-crdb(suggestion): Missing `bestStoreID == 0` guard. `pickStoreByDiversity` documents returning 0 for "no valid candidate." While currently safe because `filterAddCandidates` guarantees non-nil storeStates, `promoteNonVoterToVoter` has this guard (line 489) and this path should too for defensive consistency.

```suggestion
	if bestStoreID == 0 {
		log.KvDistribution.Warningf(ctx,
			"skipping AddVoter repair for r%d: no valid target after diversity scoring", rangeID)
		return
	}
	targetSS := re.stores[bestStoreID]
```

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+		(*existingReplicaLocalities).getScoreChangeForNewReplica)
+	if bestStoreID == 0 {
+		log.KvDistribution.Warningf(ctx,
+			"skipping AddVoter repair for r%d: no valid promotion candidates", rangeID)
+		return
+	}
+
+	// Find the existing replica state for the non-voter being promoted.
+	prevState, found := replicaStateForStore(rs, bestStoreID)
+	if !found {
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go:506]: /review-crdb(suggestion): This "replica not found" case indicates an internal invariant violation — `bestStoreID` was just returned by `candidatesToConvertFromNonVoterToVoter()` from the same `rs.replicas` data within the same single-threaded `repair()` call. Consider making this an assertion failure (matching the pattern in `updateRepairAction` line 256):

```go
err := errors.AssertionFailedf(
    "non-voter on s%d not found in replicas for r%d", bestStoreID, rangeID)
if buildutil.CrdbTestBuild {
    panic(err)
}
log.KvDistribution.Warningf(ctx,
    "skipping AddVoter repair for r%d: %v", rangeID, err)
return
```

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+			// Only repair ranges where localStoreID is the leaseholder.
+			if !isLeaseholderOnStore(rs, localStoreID) {
+				continue
+			}
+
+			switch action {
+			case AddVoter:
+				re.repairAddVoter(ctx, localStoreID, rangeID, rs)
+			default:
+				log.KvDistribution.Infof(ctx,
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: /review-crdb(suggestion): With `IncludeRepair` running on every `ComputeChanges` pass, this Info-level log fires for every range needing an unimplemented action (FinalizeAtomicReplicationChange, RemoveLearner, ReplaceDeadVoter, etc.) on every pass. Consider `VEventf` to reduce noise:

```suggestion
				log.KvDistribution.VEventf(ctx, 2,
```

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+	addChange := MakeAddReplicaChange(rangeID, rs.load, addIDAndType, addTarget)
+	rangeChange := MakePendingRangeChange(rangeID, []ReplicaChange{addChange})
+	if err := re.preCheckOnApplyReplicaChanges(rangeChange); err != nil {
+		log.KvDistribution.Warningf(ctx,
+			"skipping AddVoter repair for r%d: pre-check failed: %v", rangeID, err)
+		return
+	}
+	re.enactRepair(ctx, localStoreID, rangeChange)
+	log.KvDistribution.Infof(ctx,
+		"result(success): AddVoter repair for r%v, adding voter on s%v",
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: /review-crdb(nit): Uses `r%v` and `s%v` while `promoteNonVoterToVoter` (line 523) uses `r%d` and `s%d` for the same types. Minor inconsistency.

```suggestion
		"result(success): AddVoter repair for r%d, adding voter on s%d",
```

**Review by tbg — COMMENTED** [review]
## Review of commits after "START REVIEWING HERE"

Two commits add repair orchestration and the first concrete repair action (AddVoter with non-voter promotion) to the MMA prototype. The code is well-structured, the commit split is clean, and the core algorithms (reservoir sampling, priority iteration, constraint filtering) are correct.

**Verified correct:**
- Concurrency: `repair()` runs under `a.mu.Lock()` via `ComputeChanges`. No concurrent access issues.
- Map mutation during iteration: `repair()` snapshots range IDs into `ids` before iterating, so `enactRepair` -> `addPendingRangeChange` -> `updateRepairAction` mutations are safe.
- Repair-rebalance interaction: repair pending changes correctly prevent rebalancer from touching the same ranges.
- Reservoir sampling in `pickStoreByDiversity` is correct standard k=1 reservoir sampling.

**Strengths:**
- Clean two-commit structure: orchestration skeleton first, concrete action second.
- `diversityScorer` function type is a clean abstraction using idiomatic Go method value expressions.
- Non-voter promotion avoids unnecessary data movement.
- Thorough multi-round DSL tests demonstrating the full repair lifecycle.

**Test coverage gaps** (not blocking, but worth adding):
- Leaseholder-only filtering in `repair()` is untested — both test files always call `repair store-id=1` which is the leaseholder. Add a test where the local store is not the leaseholder to verify skipping.
- No tests for error/edge cases: no valid candidates, constraint analysis failure, pre-check failure.
- Diversity scoring with clearly unequal candidates not tested (only tied or single-candidate scenarios).

**Other notes:**
- The `mmaid` field comment on `rebalanceEnv` says "a counter for rebalanceStores calls" but `repair()` now also increments it. Worth updating.
- The `repair()` return value is discarded at `allocator_state.go:349` since `rebalanceStores()` returns the same accumulated `re.changes`. Consider removing the return or adding a comment.

---
*(made with [/review-crdb](https://github.com/cockroachdb/cockroach/blob/master/.claude/skills/review-crdb/SKILL.md))*

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: Done.

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go (hunk 2 above)
Comment (tbg) [reply pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go:460]: Done.

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go (hunk 3 above)
Comment (tbg) [reply pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go:506]: Done. Made it an `AssertionFailedf` with panic in test builds, warning in production.

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go (hunk 4 above)
Comment (tbg) [reply pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: Done.

File: pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go (hunk 5 above)
Comment (tbg) [reply pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go]: Done.

**Review by tbg — COMMENTED** [review]
> @tbg reviewed 35 files and all commit messages, made 3 comments, and resolved 16 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/165423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained.

___
*[`-- commits` line 243 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnW7D0Y4eOt2jgaXBNT:-OnW7D0Y4eOt2jgaXBNU:bporui9):*
Oh. Well I like this mmaid change. But I don't like the churn in the testdata files. Can you update the datadriven DSL directive so that it maintains an mmaid counter in the context separately in the test? Basically, make the testdata diff go away. Add a comment explaining why this is currently non-unified.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 255 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnW5JBQBz9w_x5Tvej-:-OnW5JBQBz9w_x5Tvej0:bqr76b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4658125f3543d1b4b03a94595c439726ee103b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L255)):*
> ```Go
> 	}
> 	// Remove from old bucket.
> 	if oldAction != NoRepairNeeded && oldAction != RepairPending && oldAction != 0 {
> ```

The contract about what is or isn't in `repairRanges` is a little obfuscated. Looks like `RepairPending` and `RepairSkipped` are special in that they aren't tracked. But why not, wouldn't that be more consistent and also more helpful for further metrics (not in this PR), where we will want to expose a gauge that has the count by state? I think we should explicitly track all states, except `NoRepairNeeded`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 555 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnW6B4BEOJkjiARH31e:-OnW6B4BEOJkjiARH31f:b-2old5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4658125f3543d1b4b03a94595c439726ee103b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L555)):*
> ```Go
> 		// Collect and sort range IDs, then shuffle deterministically so that
> 		// iteration is not systematically biased toward any range.
> 		idsPtr := rangeIDSlicePool.Get().(*[]roachpb.RangeID)
> ```

can you lift this pool outside the `for` loop, use `defer` to return the slices to the pool, and reset the slices at the top of each loop?

**Review by tbg — COMMENTED** [review]
> @tbg reviewed 21 files and all commit messages, made 2 comments, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/165423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained.

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 569 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnWBDH-8Ov-fu3ClL_m:-OnWBDH-8Ov-fu3ClL_n:b-76o70i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d23c04ee04926943435302d5140772eba5cc45b3/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L569)):*
> ```Go
> 			ids[i], ids[j] = ids[j], ids[i]
> 		})
> 		*idsPtr = ids // preserve any grown capacity for next iteration
> ```

this pattern seems cleaner:

```go
idsPtr := ...
ids := (*idsPtr)[:0]
defer func() {
  // Preserve any grown capacity for next iteration.
  ids = ids[:0]
  *idsPtr = ids
  rangeIDSlicePool.Put(idsPtr)
}()
for ... {
  ids = ids[:0]

  ids = append(ids, ...)
}
```

WDYT?
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go` line 581 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/165423#-OnWB3Pq93-tl-SRs62X:-OnWB3Pq93-tl-SRs62Y:b-nujorb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d23c04ee04926943435302d5140772eba5cc45b3/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_repair.go#L581)):*
> ```Go
> 			case AddVoter:
> 				re.repairAddVoter(ctx, localStoreID, rangeID, rs)
> 			case RepairSkipped, RepairPending:
> ```

can you move this into the `len(ranges) == 0` check at the top so we don't needlessly iterate over all the individual ranges in cases where we're going to not act on any?

**Review by tbg — COMMENTED** [review]
> @tbg reviewed 1 file and all commit messages, and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/165423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on tbg).

**Review by tbg — COMMENTED** [review]
> @tbg reviewed 10 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/165423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on tbg).
